### PR TITLE
[TorchScript] C++ interface of to_<backend>

### PR DESCRIPTION
**Pull request by iseeyuan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51797 [TorchScript] C++ interface of to_<backend>**

The C++ API, ```codegen_backend_module``` is added to ```to_<backend>```. Python related stuffs are decoupled in this function. It can be used from both C++ and python. 

* Tests
Python: The existing ```test_backends.py```, which calls the C++ API under the hood.
C++: The end-to-end test of ```jit.BackendTest.ToBackend``` is added in ```test_backend.cpp```. The original class definitions in this file is moved to ```test_backend_lib.cpp```

Differential Revision: [D26280518](https://our.internmc.facebook.com/intern/diff/D26280518/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26280518/)!

### CODE REVIEW

File: torch/csrc/jit/backends/backend_detail.h
Comment (iseeyuan) [reply torch/csrc/jit/backends/backend_detail.h:63]: Right, they are referenced by codegen_backend_module

File: torch/csrc/jit/backends/backend_detail.h
Comment (raziel) [inline torch/csrc/jit/backends/backend_detail.h:63]: I think you can move this to the same line (otherwise CLANFORMAT will likely complain).

**Review by raziel — COMMENTED** [review]
LGTM

I'll leave it to @SplitInfinity to approve.

File: test/cpp/jit/CMakeLists.txt
Comment (SplitInfinity) [inline test/cpp/jit/CMakeLists.txt:77]: What's the reason for this change? Is it because `jitbackend_test` includes `torch`?

File: test/cpp/jit/test_backend.cpp
Comment (SplitInfinity) [inline test/cpp/jit/test_backend.cpp:5]: Why is this header needed? I don't see any of the types from the file used here.

File: test/cpp/jit/test_backend.cpp
Comment (SplitInfinity) [inline test/cpp/jit/test_backend.cpp:6]: Why is this header needed? I'm not familiar with it so it might be used, I'm just curious.

File: test/cpp/jit/test_backend_lib.cpp
Comment (SplitInfinity) [inline test/cpp/jit/test_backend_lib.cpp:9]: As I mentioned before, removing `TORCH_API` might fix your test failures.

File: torch/csrc/jit/backends/backend_detail.h
Comment (SplitInfinity) [inline torch/csrc/jit/backends/backend_detail.h:63]: Why is `TORCH_API` no longer needed? Is it because these functions are used only in `libtorch` after moving the codegen to `backend_detail.cpp`, and no longer used in `libtorch_python`?

**Review by SplitInfinity — APPROVED** [review]
Changes look good, but please fix CI before merging.

File: torch/csrc/jit/backends/backend_init.cpp
Comment (SplitInfinity) [inline torch/csrc/jit/backends/backend_init.cpp:132]:
```suggestion
}
```

File: torch/csrc/jit/backends/backend_init.cpp
Comment (SplitInfinity) [inline torch/csrc/jit/backends/backend_init.cpp:121]: I would rename this now that it is no longer a lambda.

File: test/cpp/jit/CMakeLists.txt
Comment (iseeyuan) [reply test/cpp/jit/CMakeLists.txt:77]: Yes. I'm trying to resolve CI failures and testing if this change could help. It turns out not. I may change it back after verification.

File: test/cpp/jit/test_backend.cpp
Comment (iseeyuan) [reply test/cpp/jit/test_backend.cpp:6]: It's a header of all file, probably for ease of use, especially for generated headers? The tensor factory methods (torch::ones) are from what this header includes.

File: torch/csrc/jit/backends/backend_detail.h
Comment (raziel) [reply torch/csrc/jit/backends/backend_detail.h:63]: Yep, I originally added the TORCH_API because of the split but now shouldn't be needed.

Martin can confirm that's the case now.

Comment (iseeyuan) [reply torch/csrc/jit/backends/backend_detail.h:63]: It's referenced by codegen_backend_module, which is already marked as ```TORCH_API```

File: test/cpp/jit/test_backend.cpp
Comment (iseeyuan) [reply test/cpp/jit/test_backend.cpp:5]: Removed.

**Review by raziel — APPROVED** [review]

File: test/cpp/jit/CMakeLists.txt
Comment (malfet) [inline test/cpp/jit/CMakeLists.txt:78]: On Linux, linker would not add a dependency for shared library, unless target actually depends on symbol from that library.
One way to workaround it is to add something like:
```
+if(LINUX)
  #Update to target_link_options when CMake version can be upgraded
  target_link_libraries(test_jit PRIVATE "-Wl,--no-as-needed,$<TARGET_FILE:jitbackend_test>")
endif()
```

Comment (iseeyuan) [reply test/cpp/jit/CMakeLists.txt:77]: Thanks @malfet !

File: torch/custom_class_detail.h
Comment (raziel) [inline torch/custom_class_detail.h:138]: maybe revert this change?

Comment (iseeyuan) [reply torch/custom_class_detail.h]: Good catch. Just updated it, and also rebase to internal fbobjc/stable for easier landing.

**Pull request closed without merging**
